File: UVa/java/UVa_00401_Palindromes.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {

        Map<Character, Character> mirror = new HashMap<Character, Character>(21);
        mirror.put('A', 'A'); mirror.put('E', '3'); mirror.put('H', 'H');
        mirror.put('I', 'I'); mirror.put('J', 'L'); mirror.put('L', 'J');
        mirror.put('M', 'M'); mirror.put('O', 'O'); mirror.put('S', '2');
        mirror.put('T', 'T'); mirror.put('U', 'U'); mirror.put('V', 'V');
        mirror.put('W', 'W'); mirror.put('X', 'X'); mirror.put('Y', 'Y');
        mirror.put('Z', '5'); mirror.put('1', '1'); mirror.put('2', 'S');
        mirror.put('3', 'E'); mirror.put('5', 'Z'); mirror.put('8', '8');

        Scanner scanner = new Scanner(System.in);
        String line;
        boolean isPalindrome, isMirrored;

        do{

            line = scanner.nextLine();
            isPalindrome = isPalindrome(line);
            isMirrored = isMirrored(line, mirror);

            if(!isPalindrome && !isMirrored){
                System.out.println(line + " -- is not a palindrome.");
            }else if(isPalindrome && isMirrored){
                System.out.println(line + " -- is a mirrored palindrome.");
            }else if(isPalindrome){
                System.out.println(line + " -- is a regular palindrome.");
            }else if(isMirrored){
                System.out.println(line + " -- is a mirrored string.");
            }

            System.out.println();
        }while(scanner.hasNextLine());

    }

    public static boolean isPalindrome(String line) {
        boolean itIs = true;
        for(int i=0; i<line.length()/2 && itIs; ++i){
            itIs &= line.charAt(i) == line.charAt((line.length()-1)-i);
        }

        return itIs;
    }

    public static boolean isMirrored(String line, Map<Character, Character> mirror) {
        boolean itIs = true;
        char a, b;
        int limit = line.length()%2 == 1 ? ((line.length()/2)+1) : line.length()/2;
        for (int i=0; i<limit && itIs; ++i) {
            a = line.charAt(i);
            b = line.charAt((line.length()-1)-i);

            itIs = mirror.containsKey(a) && mirror.get(a) == b;
        }

        return itIs;
    }
}

File: UVa/kotlin/UVa_00401_Palindromes.kt
import kotlin.collections.*
import kotlin.io.*
import java.util.Scanner

val reverseMap = mapOf(
    'A' to 'A', 'E' to '3', 'H' to 'H', 'I' to 'I',
    'J' to 'L', 'L' to 'J', 'M' to 'M', 'O' to 'O',
    'S' to '2', 'T' to 'T', 'U' to 'U', 'V' to 'V',
    'W' to 'W', 'X' to 'X', 'Y' to 'Y', 'Z' to '5',
    '1' to '1', '2' to 'S', '3' to 'E', '5' to 'Z',
    '8' to '8'
)

fun main() {
    val scanner = Scanner(System.`in`)
    var isPalindrome = true
    var isMirrored = true

    while (scanner.hasNext()) {
        var line = scanner.nextLine()

        for (i in 0..<line.length) {
            isPalindrome = isPalindrome && line[i] == line[line.length - i - 1]
            isMirrored = isMirrored && line[line.length - i - 1] == reverseMap.get(line[i])

            if (!isPalindrome && !isMirrored) break
        }

        if (isPalindrome && isMirrored) {
            println("$line -- is a mirrored palindrome.")
        } else if (isMirrored) {
            println("$line -- is a mirrored string.")
        } else if (isPalindrome) {
            println("$line -- is a regular palindrome.")
        } else {
            println("$line -- is not a palindrome.")
        }
        println()
        isPalindrome = true
        isMirrored = true
    }
}
